rasa/nlu/classifiers/logistic_regression_classifier.py

```diff
@@ -0,0 +1,200 @@
+        )
+
+        # We need to use these later when saving the trained component.
+        self._model_storage = model_storage
+        self._resource = resource
+
+    def _create_X(self, messages: List[Message]) -> csr_matrix:
+        """This method creates a sparse X array that can be used for predicting."""
+        X = []
+        for e in messages:
+            # First element is sequence features, second is sentence features
+            sparse_feats = e.get_sparse_features(attribute=TEXT)[1]
+            # First element is sequence features, second is sentence features
+            dense_feats = e.get_dense_features(attribute=TEXT)[1]
+            together = hstack(
+                [
+                    csr_matrix(sparse_feats.features if sparse_feats else []),
+                    csr_matrix(dense_feats.features if dense_feats else []),
+                ]
+            )
+            X.append(together)
+        return vstack(X)
+
+    def _create_training_matrix(
+        self, training_data: TrainingData
+    ) -> Tuple[csr_matrix, List[str]]:
+        """This method creates a scikit-learn compatible (X, y) training pairs."""
+        y = []
+
+        examples = [
+            e
+            for e in training_data.intent_examples
+            if (e.get("intent") and e.get("text"))
+        ]
+
+        for e in examples:
+            y.append(e.get(INTENT))
+
+        return self._create_X(examples), y
+
+    def train(self, training_data: TrainingData) -> Resource:
+        """Train the intent classifier on a data set."""
+        X, y = self._create_training_matrix(training_data)
+        if X.shape[0] == 0:
+            logger.debug(
+                f"Cannot train '{self.__class__.__name__}'. No data was provided. "
+                f"Skipping training of the classifier."
+            )
+            return self._resource
+
+        self.clf.fit(X, y)
+        self.persist()
+
+        return self._resource
+
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> GraphComponent:
+        """Creates a new untrained component (see parent class for full docstring)."""
+        return cls(config, execution_context.node_name, model_storage, resource)
+
+    def process(self, messages: List[Message]) -> List[Message]:
+        """Return the most likely intent and its probability for a message."""
+        X = self._create_X(messages)
+        pred = self.clf.predict(X)
+        probas = self.clf.predict_proba(X)
+        for idx, message in enumerate(messages):
+            intent = {"name": pred[idx], "confidence": probas[idx].max()}
+            intents = self.clf.classes_
+            intent_info = {
+                k: v
+                for i, (k, v) in enumerate(zip(intents, probas[idx]))
+                if i < LABEL_RANKING_LENGTH
+            }
+            intent_ranking = [
```

Comment:
One issue I noticed with using this component with Moodbot is that the intent ranking seems to be unsorted. For a user text `hello`, I got the following scores:

```
[{'name': 'affirm', 'confidence': 0.018605408851111678}, {'name': 'bot_challenge', 'confidence': 0.0006403107430143258}, {'name': 'deny', 'confidence': 0.025455037690681255}, {'name': 'goodbye', 'confidence': 0.013747843491355998}, {'name': 'greet', 'confidence': 0.914782164381611}, {'name': 'mood_great', 'confidence': 0.017658700205053904}, {'name': 'mood_unhappy', 'confidence': 0.009110534637171744}]
```

This in turn leads to `FallbackClassifier` being triggered, since it will look at the difference of the first two elements, which is below the default ambiguity threshold. Therefore the probabilities need to be sorted, as e.g. [here](https://github.com/RasaHQ/rasa/blob/main/rasa/nlu/classifiers/sklearn_intent_classifier.py#L257).

Reply:
Well spotted!

Reply:
Also added a test for it.